src/platform_impl/linux/wayland/event_loop/mod.rs

```diff
@@ -128,330 +145,411 @@ impl<T: 'static> EventLoop<T> {
         Ok(event_loop)
     }
 
-    pub fn run<F>(mut self, callback: F) -> !
+    pub fn run_ondemand<F>(&mut self, mut event_handler: F) -> Result<(), RunLoopError>
     where
-        F: FnMut(Event<'_, T>, &RootEventLoopWindowTarget<T>, &mut ControlFlow) + 'static,
+        F: FnMut(Event<'_, T>, &RootEventLoopWindowTarget<T>, &mut ControlFlow),
     {
-        let exit_code = self.run_return(callback);
-        process::exit(exit_code);
+        if self.loop_running {
+            return Err(RunLoopError::AlreadyRunning);
+        }
+
+        let exit = loop {
+            match self.pump_events_with_timeout(None, &mut event_handler) {
+                PumpStatus::Exit(0) => {
+                    break Ok(());
+                }
+                PumpStatus::Exit(code) => {
+                    break Err(RunLoopError::ExitFailure(code));
+                }
+                _ => {}
+            }
+        };
+
+        // Applications aren't allowed to carry windows between separate
+        // `run_ondemand` calls but if they have only just dropped their
+        // windows we need to make sure those last requests are sent to the
+        // compositor.
+        let _ = self.roundtrip().map_err(RunLoopError::Os);
```

Comment:
We could simply `let _ = self.connection.flush();` here, no need for roundtrip actually, though, it would be nice to drain stuff out of socket.

Also, we should consider that compositor could send events even though we destroy all of the windows, like `wl_registry` events, so ideally we should create background thread when returning from `run_ondemand` to read the connection from time to time. Though, we could leave it to separate PR...

Reply:
Yeah I think I just went with what you initially suggested to have a full sync roundtrip but yeah it sounds reasonable that a flush could be enough. Did you already test that `.flush()` is enough? 

Looking at the `Drop` implementation for `Window` it actually looks like it doesn't immediately write a request to the connection so maybe flush isn't enough? Since `Drop` seems to buffer via `state.window_requests` it looks like a Window Drop actually requires an event loop iteration, followed by a flush() - so I guess a different example could be construed that would still have an issue on Wayland even with the full sync roundtrip?

Going as far as having a background thread etc to service an idle connection sounds like it would be going too far.

If we end up realizing that the wayland backend can't maintain an idle connection without a running event loop then we can just drop the connection at the end of `run_ondemand`, and re-open if `run_ondemand` is called again.

It's not like on macOS where there's a global NSApp state which makes it more awkward to ensure separate runs of run_ondemand are orthogonal.

Dropping the connection would also be the most robust solution for ensuring windows are destroyed.

Reply:
>Yeah I think I just went with what you initially suggested to have a full sync roundtrip but yeah it sounds reasonable that a flush could be enough. Did you already test that .flush() is enough?

Yes, I've tested and it was enough.